labs/passport-adapter/README.md

```diff
@@ -0,0 +1,185 @@
+      AuthenticationBindings.AUTHENTICATION_STRATEGY_EXTENSION_POINT_NAME,
+  });
+```
+
+### With Provider
+
+If you need to inject stuff (e.g. the verify function) when configure the
+strategy, you may want to provide your strategy as a provider.
+
+_Note: If you are not familiar with LoopBack provider, check the documentations
+in
+[Extending LoopBack 4](https://loopback.io/doc/en/lb4/Extending-LoopBack-4.html)_
+
+1. Create a provider for the strategy
+
+Use `passport-http` as the example again:
+
+```ts
+class PassportBasicAuthProvider implements Provider<AuthenticationStrategy> {
+  value(): AuthenticationStrategy {
+    // The code that returns the converted strategy
+  }
+}
+```
+
+The Provider should have two functions:
+
+- A function that takes in the verify callback function and returns a configured
+  basic strategy. To know more about the configuration, please check
+  [the configuration guide in module `passport-http`](https://github.com/jaredhanson/passport-http#usage-of-http-basic).
+
+- A function that applies the `StrategyAdapter` to the configured basic strategy
+  instance. Then in the `value()` function, you return the converted strategy.
+
+So a full implementation of the provider is:
+
+```ts
+import {BasicStrategy, BasicVerifyFunction} from 'passport-http';
+import {StrategyAdapter} from `@loopback/passport-adapter`;
+import {AuthenticationStrategy} from '@loopback/authentication';
+
+class PassportBasicAuthProvider implements Provider<AuthenticationStrategy> {
+  constructor(@inject(VERIFY) verifyFn: BasicVerifyFunction);
+  value(): AuthenticationStrategy {
+    const basicStrategy = this.configuredBasicStrategy(verify);
+    return this.convertToAuthStrategy(basicStrategy);
+  }
+
+  // Takes in the verify callback function and returns a configured basic strategy.
+  configuredBasicStrategy(verifyFn: BasicVerifyFunction): BasicStrategy {
+    return new BasicStrategy(verifyFn);
+  }
+
+  // Applies the `StrategyAdapter` to the configured basic strategy instance.
+  // You'd better define your strategy name as a constant, like
+  // `const AUTH_STRATEGY_NAME = 'basic'`
+  // So that you will decorate the APIs later with the same name
+  convertToAuthStrategy(basic: BasicStrategy): AuthenticationStrategy {
+    return new StrategyAdapter(basic, AUTH_STRATEGY_NAME);
+  }
+}
+```
+
+2. Register the strategy provider
+
+Register the strategy provider in your LoopBack application so that the
+authentication system can look for your strategy by name and invoke it:
+
+```ts
+// In the main file
+
+import {addExtension} from '@loopback/core';
+import {MyApplication} from '<path_to_your_app>';
+import {AuthenticationBindings} from '@loopback/authentication';
+
+const app = new MyApplication();
+
+addExtension(
+  app,
+  AuthenticationBindings.AUTHENTICATION_STRATEGY_EXTENSION_POINT_NAME,
+  PassportBasicAuthProvider,
+  {
+    namespace:
+      AuthenticationBindings.AUTHENTICATION_STRATEGY_EXTENSION_POINT_NAME,
+  },
+);
```

Comment:
I find this as not very elegant. Can we find a way how to simplify this registration step? For example, can we leverage `@extension` or `@bind` decorators to specify the extension point name in the code where we are defining `PassportBasicAuthProvider`.

@raymondfeng you are the author of `addExtension` and the extension-point/extension implementation. What is the recommended way for registering extensions? Are we missing any new helpers in `@loopback/core` to simplify this process?

Reply:
My code is out of date we do have that sugar function :)

https://github.com/strongloop/loopback-next/pull/3056#discussion_r293112596